src/providers/mssql/qgsmssqltransaction.cpp

```diff
@@ -0,0 +1,158 @@
+/***************************************************************************
+  qgsmssqltransaction.cpp - QgsMssqlTransaction
+ ---------------------
+ begin                : 11.3.2021
+ copyright            : (C) 2021 by Vincent Cloarec
+ email                : vcloarec at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmssqltransaction.h"
+
+#include "qgsmssqldatabase.h"
+
+#include "qgsexpression.h"
+#include "qgsmessagelog.h"
+
+#include <QSqlError>
+#include <QSqlQuery>
+#include <QtDebug>
+
+
+QgsMssqlTransaction::QgsMssqlTransaction( const QString &connString )
+  : QgsTransaction( connString )
+{
+}
+
+QgsMssqlTransaction::~QgsMssqlTransaction()
+{
+}
+
+
+bool QgsMssqlTransaction::executeSql( const QString &sql, QString &error, bool isDirty, const QString &name )
+{
+  QSqlDatabase &database = mConn->db();
+
+  if ( !database.isValid() )
```

Comment:
I am wondering if isValid() doesn't imply automatically that is it open()

Reply:
No, QSqlDatabase::isValid() == true if there is a valid driver assigned. To check if it is open, there is QSqlDatabase::isOpen()